stm32: ST Kit B-L4S5I-IOT01A Octospi flash support

**Is your enhancement proposal related to a problem? Please describe.**

The B-L4S5I-IOT01A board has an onboard QSPI flash, but the flash memory is not exposed in the device tree. 

**Describe the solution you'd like**

I would like the ability to use the external flash memory with the B-L4S5I-IOT01A board. 

**Describe alternatives you've considered**

I tried adding support for ospi flash to the device tree from v3.2.0, but ran into driver compatibility issues with the stm32l4 ospi hal (see below).

The [Stm32CubeL4 HAL implementation](https://github.com/zephyrproject-rtos/hal_stm32/blob/main/stm32cube/stm32l4xx/drivers/include/stm32l4xx_hal_ospi.h) does not appear to be compatible with the [OSPI flash driver](https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v3.2.0/drivers/flash/flash_stm32_ospi.c). Curiously there does appear to be a compatible implementation of stm32L4xx_hal_ospi out there (for example the [implementation used by the LoRaMac-Node](https://github.com/zephyrproject-rtos/loramac-node/blob/zephyr/src/boards/mcu/stm32/STM32L4xx_HAL_Driver/Inc/stm32l4xx_hal_ospi.h)).

### Errors from build: 

```bash
/home/brian/zephyrproject/zephyr/drivers/flash/flash_stm32_ospi.c: In function 'flash_stm32_ospi_init':
/home/brian/zephyrproject/zephyr/drivers/flash/flash_stm32_ospi.c:1766:29: error: 'OSPI_InitTypeDef' has no member named 'WrapSize'
 1766 |         dev_data->hospi.Init.WrapSize = HAL_OSPI_WRAP_NOT_SUPPORTED;
      |                             ^
/home/brian/zephyrproject/zephyr/drivers/flash/flash_stm32_ospi.c:1766:41: error: 'HAL_OSPI_WRAP_NOT_SUPPORTED' undeclared (first use in this function); did you mean 'HAL_DMA_ERROR_NOT_SUPPORTED'?
 1766 |         dev_data->hospi.Init.WrapSize = HAL_OSPI_WRAP_NOT_SUPPORTED;
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                         HAL_DMA_ERROR_NOT_SUPPORTED
/home/brian/zephyrproject/zephyr/drivers/flash/flash_stm32_ospi.c:1766:41: note: each undeclared identifier is reported only once for each function it appears in
/home/brian/zephyrproject/zephyr/drivers/flash/flash_stm32_ospi.c:1778:29: error: 'OSPI_InitTypeDef' has no member named 'Refresh'
 1778 |         dev_data->hospi.Init.Refresh = 0;
      |                             ^
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/brian/firmware/build
```

### Changes to the device tree:

```
diff --git a/dts/arm/st/l4/stm32l4r5.dtsi b/dts/arm/st/l4/stm32l4r5.dtsi
index 6bd658232a..d85ac9b546 100644
--- a/dts/arm/st/l4/stm32l4r5.dtsi
+++ b/dts/arm/st/l4/stm32l4r5.dtsi
@@ -6,6 +6,7 @@
 
 #include <mem.h>
 #include <st/l4/stm32l4.dtsi>
+#include <zephyr/dt-bindings/flash_controller/ospi.h>
 
 / {
        sram0: memory@20000000 {
@@ -246,6 +247,17 @@
                        };
                };
 
+    // 0xA000 1400 - 0xA000 13FF
+               octospi2: octospi@a0001000 {
+                       compatible = "st,stm32-ospi";
+      reg = <0xA0001000 0x400>; 
+                       interrupts = <76 0>;  
+      clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00000100>; 
+                       #address-cells = <1>;
+                       #size-cells = <0>;
+                       status = "disabled";
+               };
+
                can1: can@40006400 {
                        compatible = "st,stm32-can";
                        reg = <0x40006400 0x400>;
```

```
diff --git a/boards/arm/b_l4s5i_iot01a/b_l4s5i_iot01a.dts b/boards/arm/b_l4s5i_iot01a/b_l4s5i_iot01a.dts
index ffe1b59029..e2aa37a02e 100644
--- a/boards/arm/b_l4s5i_iot01a/b_l4s5i_iot01a.dts
+++ b/boards/arm/b_l4s5i_iot01a/b_l4s5i_iot01a.dts
@@ -205,6 +205,43 @@
        };
 };
 
+&octospi2 {
+    pinctrl-0 = <&octospim_p1_clk_pe10 &octospim_p1_ncs_pe11
+                 &octospim_p1_io0_pe12 &octospim_p1_io1_pe13
+                 &octospim_p1_io2_pe14 &octospim_p1_io3_pe15
+                 &octospim_p1_io4_pd4 &octospim_p1_io5_pd5
+                 &octospim_p1_io6_pd6 &octospim_p1_io7_pd7>;
+
+
+    pinctrl-names = "default";
+
+    status = "okay";
+
+    mx25r6435f: ospi-nor-flash@0 {
+            compatible = "st,stm32-ospi-nor";
+            reg = <0>;
+            ospi-max-frequency = <DT_FREQ_M(50)>;
+            size = <DT_SIZE_M(64)>;
+            spi-bus-width = <OSPI_QUAD_MODE>; /* access on 4 data lines */
+            data-rate = <OSPI_STR_TRANSFER>; /* access in STR */
+            status = "okay";
+    };
+
+    partitions {
+                          compatible = "fixed-partitions";
+                          #address-cells = <1>;
+                          #size-cells = <1>;
+
+                          partition@0 {
+                              label = "nor";
+                              reg = <0x00000000 DT_SIZE_M(4)>;
+                          };
+               };
+
+};
+
 &rtc {
        status = "okay";
 };
```

**Additional context**

I modeled my changes after [this pull request](https://github.com/zephyrproject-rtos/zephyr/pull/39770) by @FRASTM  though the st,stm32-ospi-nor binding has been updated to allow for quadspi mode since.

--- a/boards/arm/b_l4s5i_iot01a/b_l4s5i_iot01a.dts
+++ b/boards/arm/b_l4s5i_iot01a/b_l4s5i_iot01a.dts
@@ -18,6 +18,8 @@
 		zephyr,shell-uart = &usart1;
 		zephyr,sram = &sram0;
 		zephyr,flash = &flash0;
+		zephyr,code-partition = &slot0_partition;
+		zephyr,flash-controller = &mx25r6435f;
 	};
 
 	leds {
@@ -45,6 +47,7 @@
 		sw0 = &user_button;
 		watchdog0 = &iwdg;
 		accel0 = &lsm6dsl;
+		spi-flash0 = &mx25r6435f;
 	};
 };
 
@@ -227,5 +230,48 @@ zephyr_udc0: &usbotg_fs {
 };
 
 &rng {
+	status = "okay";
+};
+
+&octospi1 {
+	pinctrl-0 = <&octospim_p1_clk_pe10 &octospim_p1_ncs_pe11
+		     &octospim_p1_io0_pe12 &octospim_p1_io1_pe13
+		     &octospim_p1_io2_pe14 &octospim_p1_io3_pe15>;
+	pinctrl-names = "default";
+
+	dmas = <&dma1 0 40 0x480>; /* request 40 for OCTOSPI1 */
+	dma-names = "tx_rx";
+
+	status = "okay";
+
+	mx25r6435f: ospi-nor-flash@0 {
+		compatible = "st,stm32-ospi-nor";
+		reg = <0>;
+		ospi-max-frequency = <DT_FREQ_M(26)>; /* for Voltage Range 2 */
+		size = <DT_SIZE_M(8)>; /* 64 Megabits = 8 Megabytes */
+		spi-bus-width = <OSPI_QUAD_MODE>;
+		data-rate = <OSPI_STR_TRANSFER>;
+		writeoc="PP_1_4_4";
+
+		status = "okay";
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			store_partition: partition@000 {
+				label = "store";
+				reg = <0x00000000 DT_SIZE_M(8)>;
+			};
+		};
+	};
+};
+
+&dma1 {
+	status = "okay";
+};
+
+&dmamux1 {
 	status = "okay";
 };

--- a/drivers/flash/flash_stm32_ospi.c
+++ b/drivers/flash/flash_stm32_ospi.c
@@ -196,6 +196,8 @@ static int ospi_read_access(const struct device *dev, OSPI_RegularCmdTypeDef *cm
 	struct flash_stm32_ospi_data *dev_data = dev->data;
 	HAL_StatusTypeDef hal_ret;
 
+	LOG_DBG("Instruction 0x%x", cmd->Instruction);
+
 	cmd->NbData = size;
 
 	dev_data->cmd_status = 0;
@@ -1937,7 +1939,9 @@ static int flash_stm32_ospi_init(const struct device *dev)
 	dev_data->hospi.Init.ChipSelectHighTime = 2;
 	dev_data->hospi.Init.FreeRunningClock = HAL_OSPI_FREERUNCLK_DISABLE;
 	dev_data->hospi.Init.ClockMode = HAL_OSPI_CLOCK_MODE_0;
+#if defined(OCTOSPI_DCR2_WRAPSIZE)
 	dev_data->hospi.Init.WrapSize = HAL_OSPI_WRAP_NOT_SUPPORTED;
+#endif /* OCTOSPI_DCR2_WRAPSIZE */
 	dev_data->hospi.Init.SampleShifting = HAL_OSPI_SAMPLE_SHIFTING_NONE;
 	/* STR mode else Macronix for DTR mode */
 	if (dev_cfg->data_rate == OSPI_DTR_TRANSFER) {
@@ -1949,14 +1953,16 @@ static int flash_stm32_ospi_init(const struct device *dev)
 	}
 	dev_data->hospi.Init.ChipSelectBoundary = 0;
 	dev_data->hospi.Init.DelayBlockBypass = HAL_OSPI_DELAY_BLOCK_USED;
+#if defined(OCTOSPI_DCR4_REFRESH)
 	dev_data->hospi.Init.Refresh = 0;
+#endif /* OCTOSPI_DCR4_REFRESH */
 
 	if (HAL_OSPI_Init(&dev_data->hospi) != HAL_OK) {
 		LOG_ERR("OSPI Init failed");
 		return -EIO;
 	}
 
-#if defined(CONFIG_SOC_SERIES_STM32U5X)
+#if defined(OCTOSPIM)
 	/* OCTOSPI I/O manager init Function */
 	OSPIM_CfgTypeDef ospi_mgr_cfg = {0};
 
@@ -1973,15 +1979,15 @@ static int flash_stm32_ospi_init(const struct device *dev)
 		ospi_mgr_cfg.IOLowPort = HAL_OSPIM_IOPORT_2_LOW;
 		ospi_mgr_cfg.IOHighPort = HAL_OSPIM_IOPORT_2_HIGH;
 	}
-
+#if defined(OCTOSPIM_CR_MUXEN)
 	ospi_mgr_cfg.Req2AckTime = 1;
-
+#endif /* OCTOSPIM_CR_MUXEN */
 	if (HAL_OSPIM_Config(&dev_data->hospi, &ospi_mgr_cfg,
 		HAL_OSPI_TIMEOUT_DEFAULT_VALUE) != HAL_OK) {
 		LOG_ERR("OSPI M config failed");
 		return -EIO;
 	}
-
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
 	/* OCTOSPI2 delay block init Function */
 	HAL_OSPI_DLYB_CfgTypeDef ospi_delay_block_cfg = {0};
 
@@ -1992,6 +1998,7 @@ static int flash_stm32_ospi_init(const struct device *dev)
 		return -EIO;
 	}
 #endif /* CONFIG_SOC_SERIES_STM32U5X */
+#endif /* OCTOSPIM */
 
 	/* Reset NOR flash memory : still with the SPI/STR config for the NOR */
 	if (stm32_ospi_mem_reset(dev) != 0) {

--- a/dts/arm/st/l4/stm32l4r5.dtsi
+++ b/dts/arm/st/l4/stm32l4r5.dtsi
@@ -6,6 +6,9 @@
 
 #include <mem.h>
 #include <st/l4/stm32l4.dtsi>
+#include <zephyr/dt-bindings/flash_controller/ospi.h>
+
+/delete-node/ &quadspi;
 
 / {
 	sram0: memory@20000000 {
@@ -309,6 +312,34 @@
 		adc1: adc@50040000 {
 			has-temp-channel;
 			has-vref-channel;
+		};
+
+		octospi1: octospi@a0001000 {
+			compatible = "st,stm32-ospi";
+			reg = <0xa0001000 0x400>;
+			interrupts = <71 0>;
+			clock-names = "ospix", "ospi-ker", "ospi-mgr";
+			clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00000100>,
+				<&rcc STM32_SRC_SYSCLK OSPI_SEL(0)>,
+				<&rcc STM32_CLOCK_BUS_AHB2 0x00100000>;
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
+		};
+
+		octospi2: octospi@a0001400 {
+			compatible = "st,stm32-ospi";
+			reg = <0xa0001400 0x400>;
+			interrupts = <76 0>;
+			clock-names = "ospix", "ospi-ker", "ospi-mgr";
+			clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00000200>,
+				<&rcc STM32_SRC_SYSCLK OSPI_SEL(0)>,
+				<&rcc STM32_CLOCK_BUS_AHB2 0x00100000>;
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "disabled";
 		};
 	};
 

--- a/samples/drivers/jesd216/boards/b_l4s5i_iot01a.conf
+++ b/samples/drivers/jesd216/boards/b_l4s5i_iot01a.conf
@@ -1,0 +1,8 @@
+#
+# Copyright (c) 2022 STMicroelectronics
+#
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_FLASH_STM32_OSPI=y
+CONFIG_SPI_NOR_SFDP_RUNTIME=y
+CONFIG_SPI_NOR=n

